src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -118100,6 +117987,414 @@ static void M1(C c, string? s)
+}
+
+class C : B, IEquatable<string>
+{
+    bool IEquatable<string>.Equals(string? s) => throw null!;
+
+    static void M1(C c, string? s)
+    {
+        _ = c.Equals(s)
+            ? s.ToString()
+            : s.ToString(); // 1
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (18,15): warning CS8602: Dereference of a possibly null reference.
+                //             : s.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(18, 15));
+        }
+
+        [Fact, WorkItem(40577, "https://github.com/dotnet/roslyn/issues/40577")]
+        public void IEquatableEqualsMethod_ImplicitImplInBaseType_ExplicitImplInDerivedType_02()
+        {
+            var source = @"
+using System;
+
+#nullable enable
+class B : IEquatable<string>
+{
+    public virtual bool Equals(string? s) => throw null!;
+}
+
+class C : B, IEquatable<string>
+{
+    bool IEquatable<string>.Equals(string? s) => throw null!;
+    public override bool Equals(string? s) => throw null!;
+
+    static void M1(C c, string? s)
+    {
+        _ = c.Equals(s)
+            ? s.ToString()
+            : s.ToString(); // 1
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (19,15): warning CS8602: Dereference of a possibly null reference.
+                //             : s.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(19, 15));
+        }
+
+        [Fact, WorkItem(40577, "https://github.com/dotnet/roslyn/issues/40577")]
+        public void IEquatableEqualsMethod_ExplicitImplInDerivedType()
+        {
+            var source = @"
+using System;
+
+#nullable enable
+class B
+{
+    public bool Equals(string? s) => throw null!;
+}
+
+class C : B, IEquatable<string>
+{
+    bool IEquatable<string>.Equals(string? s) => throw null!;
+
+    static void M1(C c, string? s)
+    {
+        _ = c.Equals(s)
+            ? s.ToString() // 1
+            : s.ToString(); // 2
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (17,15): warning CS8602: Dereference of a possibly null reference.
+                //             ? s.ToString() // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(17, 15),
+                // (18,15): warning CS8602: Dereference of a possibly null reference.
+                //             : s.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(18, 15));
+        }
+
+        [Fact, WorkItem(40577, "https://github.com/dotnet/roslyn/issues/40577")]
+        public void IEquatableEqualsMethod_Override()
+        {
+            var source = @"
+using System;
+
+#nullable enable
+class B : IEquatable<string>
+{
+    public virtual bool Equals(string? s) => throw null!;
+}
```

Comment:
>: s.ToString(); // 1 [](start = 12, length = 20)

It is fine to keep this test, but the interesting scenario is when a class implements I1 and we call Equals on that class.

Reply:
Done